core/threaded/scheduler_QS.c

```diff
@@ -0,0 +1,326 @@
+    LF_PRINT_DEBUG("Scheduler: New number of idle workers: %u.",
+                _lf_sched_instance->_lf_sched_number_of_idle_workers);
+    
+    if (workers_to_awaken > 1) {
+        // Notify all the workers except the worker thread that has called this
+        // function.
+        // lf_mutex_lock(&_lf_sched_instance->_lf_sched_semaphore->mutex);
+        // lf_cond_broadcast(&_lf_sched_instance->_lf_sched_semaphore->cond);
+        // lf_mutex_unlock(&_lf_sched_instance->_lf_sched_semaphore->mutex);
+        lf_cond_broadcast(&_lf_sched_instance->_lf_sched_array_of_conds[0]);
+    }
+    
+}
+
+/**
+ * @brief Signal all worker threads that it is time to stop.
+ *
+ */
+void _lf_sched_signal_stop() {
+    _lf_sched_instance->_lf_sched_should_stop = true;
+}
+
+/**
+ * @brief Wait until the scheduler assigns work.
+ *
+ * If the calling worker thread is the last to become idle, it will call on the
+ * scheduler to distribute work. Otherwise, it will wait on
+ * '_lf_sched_instance->_lf_sched_semaphore'.
+ *
+ * @param worker_number The worker number of the worker thread asking for work
+ * to be assigned to it.
+ */
+void _lf_sched_wait_for_work(size_t worker_number) {
+    // Increment the number of idle workers by 1 and check if this is the last
+    // worker thread to become idle.
+    LF_PRINT_DEBUG(
+        "Scheduler: Worker %d is trying to acquire the scheduling "
+        "semaphore.",
+        worker_number);
+    lf_mutex_lock(&mutex);
+    if (++_lf_sched_instance->_lf_sched_number_of_idle_workers ==
+        _lf_sched_instance->_lf_sched_number_of_workers) {
+        // Last thread to go idle
+        LF_PRINT_DEBUG("Scheduler: Worker %d is the last idle thread.",
+                    worker_number);
+
+        // Nothing more happening at this tag.
+        LF_PRINT_DEBUG("Scheduler: Advancing tag.");
+        // This worker thread will take charge of advancing tag.
+        if (_lf_sched_advance_tag_locked()) {
+            LF_PRINT_DEBUG("Scheduler: Reached stop tag.");
+            _lf_sched_signal_stop();
+        }
+
+        // Reset all the PCs to 0.
+        for (int w = 0; w < _lf_sched_instance->_lf_sched_number_of_workers; w++) {
+            _lf_sched_instance->pc[w] = 0;
+        }
+        // Reset all the semaphores.
+        for (int i = 0; i < num_semaphores; i++) {
+            _lf_sched_instance->semaphores[i]->count = 0 ;
+        }
+
+        
+        _lf_sched_notify_workers();
+    } else {
+        // Not the last thread to become idle. Wait for work to be released.
+        // Wait for the last thread to signal the condition variable.
+        // lf_mutex_lock(&_lf_sched_instance->_lf_sched_semaphore->mutex);
+        // lf_cond_wait(&_lf_sched_instance->_lf_sched_semaphore->cond, &_lf_sched_instance->_lf_sched_semaphore->mutex);
+        // lf_mutex_unlock(&_lf_sched_instance->_lf_sched_semaphore->mutex);
+        lf_cond_wait(&_lf_sched_instance->_lf_sched_array_of_conds[0], &mutex);
```

Comment:
Should this index be the worker number rather than 0?

Reply:
Although the field is called `_lf_sched_array_of_conds`, the current implementation only uses one condition variable to coordinate time advancement. An array is used here so that a future scheduler implementation that needs multiple condition variables can still use this field.

Reply:
A comment in the code about this would be helpful.